### PR TITLE
Implement WQuickForeignToplevelManagerV1

### PR DESCRIPTION
**Pull request by justforlxz**

### CODE REVIEW

File: src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        auto handle = std::make_shared<QWForeignToplevelHandleV1>(
+            QWForeignToplevelHandleV1::create(manager));
+        surfaces.insert({surface, handle});
+        updateSurfaceData(surface);
+    }
+
+    void remove(WXdgSurface *surface) { surfaces.erase(surface); }
+
+    void surfaceOutputEnter(WXdgSurface *surface, WOutput *output) {
+        auto handle = surfaces[surface];
```
Comment (justforlxz) [inline src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp:69]: Should I check that the surface still exists?

File: src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+#include <map>
+#include <memory>
+
+QW_USE_NAMESPACE
+
+WAYLIB_SERVER_BEGIN_NAMESPACE
+
+class WQuickForeignToplevelManagerV1Private : public WObjectPrivate {
+  public:
```
Comment (zccrs) [inline src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp:20]: no space

File: src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    WQuickForeignToplevelManagerV1Private(WQuickForeignToplevelManagerV1 *qq)
+        : WObjectPrivate(qq) {}
+    ~WQuickForeignToplevelManagerV1Private() = default;
+
+    void updateSurfaceData(WXdgSurface *surface) {
```
Comment (zccrs) [inline src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp:25]: initSurface?

File: src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        handle->outputLeave(output->handle());
+    }
+
+    W_DECLARE_PUBLIC(WQuickForeignToplevelManagerV1)
+
+    QWForeignToplevelManagerV1 *manager = nullptr;
+    std::map<WXdgSurface *, std::shared_ptr<QWForeignToplevelHandleV1>> surfaces;
+};
+
+WQuickForeignToplevelManagerV1::WQuickForeignToplevelManagerV1(QObject *parent)
```
Comment (zccrs) [inline src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp:83]:
````
XXX:XXX()
    : PPPP
    , MMM
````

File: src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            });
+    }
+
+    void add(WXdgSurface *surface) {
+        auto handle = std::shared_ptr<QWForeignToplevelHandleV1>(QWForeignToplevelHandleV1::create(manager));
+        surfaces.insert({surface, handle});
+        updateSurfaceData(surface);
+    }
+
+    void remove(WXdgSurface *surface) { surfaces.erase(surface); }
```
Comment (zccrs) [inline src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp:65]: Where do disconnect signals for this surface?

File: src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    W_D(WQuickForeignToplevelManagerV1);
+    d->add(surface);
+}
+
+void WQuickForeignToplevelManagerV1::remove(WXdgSurface *surface) {
+    W_D(WQuickForeignToplevelManagerV1);
+    d->remove(surface);
+}
+
+void WQuickForeignToplevelManagerV1::surfaceOutputEnter(WXdgSurface *surface,
```
Comment (zccrs) [inline src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp:97]: Can using signal from WSurface? See `WSurface::enterOutput`

File: src/server/kernel/wsurface.h [hunk 7]
```diff
@@ -83,6 +83,8 @@ public Q_SLOTS:
     void hasSubsurfaceChanged();
     void newSubsurface(WSurface *subsurface);
     void preferredBufferScaleChanged();
+    void enterOutputChanged(WOutput *output);
```
Comment (zccrs) [inline src/server/kernel/wsurface.h:86]: Why suffix is `changed`? Who will change?

File: src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp (hunk 1 above)
Comment (zccrs) [reply src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp:87]: You can assert it.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
+// Copyright (C) 2023 Dingyuan Zhang <lxz@mkacg.com>.
```
Comment (zccrs) [inline src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp:1]: `zhangdingyuan@uniontech.com`

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

**Pull request merged**